### PR TITLE
feat: form validation for document generator

### PR DESCRIPTION
**Pull request by Laurin-W**
Add per-field validation feedback from the validation rule results. Show error, warning, and info messages.
Needs to wait for https://github.com/inrupt/client-identifier-helper/pull/71 to be merged.

### CODE REVIEW

File: src/components/VerboseTextField.tsx [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+
+const StyledTextField = styled(TextField)<FormTextFieldProps>(({ theme }) => ({
+  "& .Mui-error": {
+    color: theme.palette.error.main,
+  },
+  "& .Mui-warning": {
+    color: theme.palette.warning.main,
+  },
+  "& .Mui-info": {
+    color: theme.palette.info.dark,
```
Comment (NSeydoux) [inline src/components/VerboseTextField.tsx:48]: Out of curiosity, why use the `dark` theme only for `info` and `main` for the other states?

File: src/components/VerboseTextField.tsx [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+    | "warning"
+    | "info"
+    | "success"
+    | "primary"
+    | undefined;
+  if (!state?.statusValue) {
+    labelColor = "primary";
+  } else {
+    labelColor = state.statusValue;
+  }
```
Comment (NSeydoux) [inline src/components/VerboseTextField.tsx:83]:
```suggestion
  const labelColor = state?.statusValue 
    ? labelColor = state.statusValue 
    : labelColor = "primary";
```
Declaring the value as const and immediately setting its value prevents from having to declare the type.

File: src/components/VerboseTextFieldArray.tsx [hunk 3]
```diff
@@ -24,15 +24,16 @@ import {
...
   TextFieldProps,
 } from "@mui/material";
+import VerboseTextField, { VerboseFieldState } from "./VerboseTextField";
 
 export type VerboseFieldArrayRenderProps = TextFieldProps & {
   /** Label for the individual text fields */
   fieldLabel: string;
   description: string | string[];
   values: string[];
+  fieldStates?: (VerboseFieldState | undefined)[];
```
Comment (NSeydoux) [inline src/components/VerboseTextFieldArray.tsx:36]: Would typing this as `VerboseFieldState[]` be sufficient, provided it is optional?

File: src/components/VerboseTextFieldArray.tsx [hunk 4]
```diff
@@ -77,13 +79,15 @@ export default function VerboseTextFieldArray(
             // eslint-disable-next-line react/no-array-index-key
             <Grid container item key={index} spacing={1}>
               <Grid item sx={{ flexGrow: 1 }}>
-                <TextField
+                <VerboseTextField
                   // eslint-disable-next-line react/jsx-props-no-spreading
```
Comment (NSeydoux) [inline src/components/VerboseTextFieldArray.tsx:83]: Usually, when a linter rule is worth ignoring, it is good to add a short comment justifying why the rule doesn't make sense in this particular context.

File: src/lib/generatorFormParameters.ts [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+  policyUri: string;
+  contact: string;
+  applicationType: "web" | "native";
+  requireAuthTime?: boolean;
+  defaultMaxAge?: number;
+};
+
+/**
+ * Return the FormParameters typed key for the string key or `undefined` if not present.
+ * If the key has the form <key name>.<array field index>, the second return value is the index.
```
Comment (NSeydoux) [inline src/lib/generatorFormParameters.ts:39]: Having such logic embedded within a parameter doesn't sound really good. Why is this necessary? I see the key is set from an event being dispatched when the form leaves focus. If parsing is required at all, it should happen then, at the top level of the call chain, and the values from there on of the key name and index should be clearly separate.

File: src/lib/generatorFormParameters.ts [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+  key: string
+): [keyof FormParameters | undefined, number | undefined] {
+  const [fieldName, index] = key.split(".");
+  if (index !== undefined && !Number.isInteger(Number(index))) {
+    return [undefined, undefined];
+  }
+
+  let field: keyof FormParameters | undefined;
+
+  switch (fieldName) {
```
Comment (NSeydoux) [inline src/lib/generatorFormParameters.ts:56]: Splitting the key between an explicit name and index will allow stricter type checking on the field name, which could be directly typed as `keyOf FormParameter`. This means the default case should not happen, and an error should be thrown otherwise.

**Review by NSeydoux — CHANGES_REQUESTED** [review]

File: src/components/VerboseTextField.tsx (hunk 1 above)
Comment (Laurin-W) [reply src/components/VerboseTextField.tsx:48]: image
image

I found the info.main color one a bit harder to read contrast-wise so I decided to go with the dark theme for this one.

Comment (Laurin-W) [reply src/components/VerboseTextField.tsx:48]: Should I change the theme instead?

File: src/components/VerboseTextFieldArray.tsx (hunk 3 above)
Comment (Laurin-W) [reply src/components/VerboseTextFieldArray.tsx:36]: For a `VerboseTextField`, having a state value is optional (no helper text and no color displayed). As the `fieldStates` are to capture the state for each `VerboseTextField` within the `TextFieldArray`, I thought this might be more consistent.

File: src/lib/generatorFormParameters.ts [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+};
+
+/**
+ * Return the FormParameters typed key for the string key.
+ * Throws an error if the key is not a member of FormParameters .
+ *
+ * @param key The FormParameters key as string.
+ * @returns FormParameters key
+ */
+export function getFormParametersKey(key: string): keyof FormParameters {
```
Comment (ThisIsMissEm) [inline src/lib/generatorFormParameters.ts:44]: A nicer way to do this would be to use a const object to define the mapping & then handle the not-a-member case.

File: src/lib/helperFunctions.ts [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+// OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+// SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+//
+
+export const underScoreToCamelCase = (str: string) => {
+  const parts = str.split("_");
+  const partsCaps = parts
+    .slice(1)
+    .map((part) => part.substring(0, 1).toUpperCase() + part.substring(1));
+  return [parts[0], partsCaps].join("");
```
Comment (ThisIsMissEm) [inline src/lib/helperFunctions.ts:27]: You could also write this using the index argument of `map`, e.g.,

```
str.split("_")
  .map((part, idx) => idx === 0 ? part : part.substring(0, 1).toUpperCase() + part.substring(1))
  .join("")
```

File: src/lib/helperFunctions.ts [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+};
+
+export const statusToNumber = (
```
Comment (ThisIsMissEm) [inline src/lib/helperFunctions.ts:30]: Ditto here you could use just an object instead of a function call

File: src/lib/validateLocalDocument.ts [hunk 10]
```diff
@@ -47,3 +48,26 @@ export default async function validateLocalDocument(
...
+ * @param validationFieldName The name of the field name in camelCase notation.
+ * @param rules the rules to validate against
+ * @returns an array of ValidationResult
+ */
+export async function validateField(
+  jsonDocument: string | object,
+  validationFieldName: string,
+  rules: ValidationRule[]
+) {
+  const results = validateLocalDocument(jsonDocument, rules);
```
Comment (ThisIsMissEm) [inline src/lib/validateLocalDocument.ts:65]: Better to lift the await up to here, instead of doing it in the next line, as you're not actually wanting to fork the promise chain.

**Review by ThisIsMissEm — COMMENTED** [review]
Looks okay, but also be careful with this getting too custom. There's also some parts that could just be object lookups.

File: src/components/VerboseTextField.tsx (hunk 1 above)
Comment (NSeydoux) [reply src/components/VerboseTextField.tsx:48]: I would say so yes, maybe run this by Kyra I'm not sure what the best practice is in this case.

Comment (Laurin-W) [reply src/components/VerboseTextField.tsx:48]: I'll catch up with Kyra and open a separate PR for this.

**Review by NSeydoux — APPROVED** [review]
